src/harness/sas_objects.py

```diff
@@ -0,0 +1,387 @@
+        cbsd_ids = self.testcase.assertRegistered(registration_requests)
+
+        #Make one CBSD object per successful grant request
+        for index, grant_request in enumerate(grant_requests):
+            grant_request["grantRequest"][0]["cbsdId"] = cbsd_ids[index]
+            grant_response = self.testcase._sas.Grant(grant_request, self.ssl_cert, self.ssl_key)[
+                'grantResponse'][0]
+            if grant_response['response']['responseCode'] == 0:
+                cbsdobject = Cbsd(cbsd_ids[index], registration_requests[index],
+                                  [grant_response['grantId']], [grant_request])
+                self.cbsd_objects.update(
+                    {grant_response['cbsdId']: cbsdobject})
+
+    def getCbsdObjectById(self, cbsd_id):
+        return self.cbsd_objects[cbsd_id]
+
+    def heartbeatRequestForAllGrants(self):
+        """
+            This function performs heartbeat request for all grants
+        """
+        for cbsd_object_item in self.cbsd_objects.values():
+            heartbeat_requests = cbsd_object_item.constructHeartbeatRequestForAllActiveGrants()
+            for heartbeat_request in heartbeat_requests:
+                heartbeat_response = self.testcase._sas.Heartbeat(
+                    heartbeat_request, self.ssl_cert, self.ssl_key)['heartbeatResponse'][0]
+                grant_object = cbsd_object_item.grant_objects[heartbeat_response['grantId']]
+                if heartbeat_response['response']['responseCode'] == 0:
+                    transmit_expire_time = datetime.strptime(heartbeat_response['transmitExpireTime'],
+                                                             '%Y-%m-%dT%H:%M:%SZ')
+                    self.testcase.assertLess(datetime.utcnow(), transmit_expire_time)
+                    self.testcase.assertLessEqual(
+                        (transmit_expire_time - datetime.utcnow()).total_seconds(), 240)
+                    grant_object.setState(Grant.AUTHORIZED)
+                else:
+                    self.mapResponseCodeToGrantState(
+                        heartbeat_response['response']['responseCode'], grant_object)
+
+    def mapResponseCodeToGrantState(self, heartbeat_response_code, grant_object):
+        """
+            This function maps the heartbeat response code to the grant state
+        """
+        if (heartbeat_response_code == 502):
+            grant_object.setState(Grant.OUT_OF_SYNC)
+        elif (heartbeat_response_code == 500):
+            grant_object.setState(Grant.TERMINATED)
+        elif (heartbeat_response_code == 501):
+            grant_object.setState(Grant.SUSPENDED)
+        else:
+            assert(not "Invalid grant response code")
+
+
+    def performHeartbeatAndUpdateGrants(self):
+        """
+            This function construct heartbeat request for all active grants and updates the grants
+        """
+        for cbsd_object_item in self.cbsd_objects.values():
```

Comment:
You need to HB for all Grants at the same time, not once per CBSD/Grant pair. The current implementation is completely antithetical to the concept of a domain proxy.

Reply:
Fixed.